tests/settings/SettingsTest.cpp

```diff
@@ -0,0 +1,263 @@
+    CPPUNIT_ASSERT_DOUBLES_EQUAL(3.3, float_matrix.m[2][0], DELTA);
+    CPPUNIT_ASSERT_DOUBLES_EQUAL(2.0, float_matrix.m[0][1], DELTA);
+    CPPUNIT_ASSERT_DOUBLES_EQUAL(3.0, float_matrix.m[1][1], DELTA);
+    CPPUNIT_ASSERT_DOUBLES_EQUAL(1.0, float_matrix.m[2][1], DELTA);
+    CPPUNIT_ASSERT_DOUBLES_EQUAL(3.0, float_matrix.m[0][2], DELTA);
+    CPPUNIT_ASSERT_DOUBLES_EQUAL(1.0, float_matrix.m[1][2], DELTA);
+    CPPUNIT_ASSERT_DOUBLES_EQUAL(2.0, float_matrix.m[2][2], DELTA);
+}
+
+void SettingsTest::addSettingVectorTest()
+{
+    settings.add("test_setting", "[0, 1, 1,2, 3 , 5,  8,13]");
+    std::vector<int> vector_int = settings.get<std::vector<int>>("test_setting");
+    std::vector<int> ground_truth = {0, 1, 1, 2, 3, 5, 8, 13};
+    CPPUNIT_ASSERT_EQUAL(ground_truth.size(), vector_int.size());
+    for (size_t i = 0; i < ground_truth.size(); i++)
+    {
+        CPPUNIT_ASSERT_EQUAL(ground_truth[i], vector_int[i]);
+    }
+
+}
+
+void SettingsTest::overwriteSettingTest()
+{
+    settings.add("test_setting", "P");
+    settings.add("test_setting", "NP");
+    CPPUNIT_ASSERT_MESSAGE("When overriding a setting, the original value was not changed.",
+                           settings.get<std::string>("test_setting") != std::string("P"));
+    CPPUNIT_ASSERT_EQUAL(std::string("NP"), settings.get<std::string>("test_setting"));
+}
+
+void SettingsTest::inheritanceTest()
+{
+    std::shared_ptr<Slice> current_slice = std::make_shared<Slice>(0);
+    Application::getInstance().current_slice = current_slice.get();
+
+    const std::string value = "A nuclear explosion would be a disaster.";
+    Settings parent;
+    parent.add("test_setting", value);
+    settings.setParent(&parent);
+
+    CPPUNIT_ASSERT_EQUAL(value, settings.get<std::string>("test_setting"));
+
+    const std::string override_value = "It's quick, it's easy and it's free: Pouring river water in your socks.";
+    settings.add("test_setting", override_value);
+    CPPUNIT_ASSERT_EQUAL_MESSAGE("The new value overrides the one from the parent.",
+                                 override_value, settings.get<std::string>("test_setting"));
+}
+
+void SettingsTest::limitToExtruderTest()
+{
+    std::shared_ptr<Slice> current_slice = std::make_shared<Slice>(0);
+    Application::getInstance().current_slice = current_slice.get();
+    current_slice->scene.extruders.emplace_back(0, nullptr);
+    current_slice->scene.extruders.emplace_back(1, nullptr);
+    current_slice->scene.extruders.emplace_back(2, nullptr);
+
+    //Add a setting to the extruder this is limiting to.
+    const std::string limit_extruder_value = "I was gonna tell a time travelling joke but you didn't like it.";
+    current_slice->scene.extruders[2].settings.add("test_setting", limit_extruder_value);
+    current_slice->scene.limit_to_extruder.emplace("test_setting", &current_slice->scene.extruders[2]);
+
+    CPPUNIT_ASSERT_EQUAL(limit_extruder_value, settings.get<std::string>("test_setting"));
```

Comment:
This test doesn't seem to test the limit_to_extruder functionality properly.

I think you should put different values for the same settings on each extruder and request the setting from a different extruder than the one limited to rather than from the `settings` object.

Reply:
I don't really understand how `settings` relates to `current_slice->scene`, though.